rfcs/20180918-functions-not-sessions-20.md

```diff
@@ -0,0 +1,1002 @@
+
+with tf.Session() as sess:
+  sess.run(init_op)
+  val = sess.run(read)
+  print(val) # Will print 1.0, the assign is ignored
+  val = sess.run([read, assign_op])[0]
+  print(val)   # Non-deterministically prints 1.0 or 2.0,
+```
+
+
+The output here is not deterministic, since `val` may evaluate to either 1.0 or 2.0 depending on whether the runtime happened to execute `assign_op` before `y` or not. `tf.control_dependencies` is a mechanism provided to add annotations at graph construction time to influence graph execution. The TensorFlow user, a Python programmer, is thus forced to think about two execution models - TensorFlow graphs and the Python interpreter. To eliminate this cognitive load, `defun` will automatically insert control dependencies to ensure that (1) operations that produce or consume a given `DT_RESOURCE` tensor and (2) operations that are marked stateful (`REGISTER_OP(...).SetIsStateful()`) follow graph construction order. Thus:
+
+
+```python
+v = tf.Variable(1.0)
+@tf.defun
+def f():
+  v.assign(2.0)
+  return v.read_value()
+
+print(f()) # Always prints 2.0.
+```
+
+
+Note that the intention here is to avoid _observable_ differences from program order. For example:
+
+
+```python
+a = tf.Variable(1.0)
+b = tf.Variable(1.0)
+@tf.defun
+def f():
+  a.assign(2.0)
+  b.assign(3.0)
+  return a + b
+print(f())
+```
+
+
+Will always print 5.0 since the assignments will occur before the read. However, there is no guaranteed ordering between the assignment of `a` and `b` (as any difference in that is not observable).
+
+A preview of this implemented in `tf.contrib.eager.defun` today (using <code>[AutomaticControlDependencies](https://github.com/tensorflow/tensorflow/blob/2f886d17f1990da418366bd093a09fb01fe5e777/tensorflow/python/eager/function.py#L1800)</code>).
+
+
+### Functions that create state
+
+In the above code, no `tf.Variable` objects are created inside a `tf.defun` decorated function. This is makes it clear that the code will have the same semantics once wrapped.
+
+Note that if the function naturally creates state only on the first trace, all is well:
+
+
+```python
+v = None
+
+@tf.defun
+def f(x):
+  global v
+  if v is None:
+    v = tf.Variable(1.0)
+  return tf.cast(x, tf.float32) + v
+
+f(tf.constant(1, dtype=tf.float32)) # Creates the variable, returns 2.0
+f(tf.constant(2, dtype=tf.int32))   # Reuses the variable, returns 3.0
+```
+
+
+To support this `defun` imposes some requirements on the decorated function:
+
+
+
+1.  State (like `tf.Variable` objects) are only created the first time the function `f` is called. \
+How that is accomplished is left up to the implementation of `f`. \
+If any variables are created in the first execution of `f`, then `@tf.defun` will trace `f` a second time in order to record the behavior that will be used from then on. No variables may be created during that second trace, or any other trace after that (due to different dtypes, shapes, or non-tensor arguments).
```

Comment:
I'm a bit confused by this sentence: what do you mean by "to record the behavior that will be used from then on"? Are you referring to the automatic addition of control dependencies for stateful ops?
Also, it seems a bit dangerous to me that `f` may be called once or twice depending on whether or not it creates variables. Can't you just "record the behavior" the first time round? Is there no way around this optional double-run? I'm sure tons of people won't expect this and their code will end up doing something twice when it should have been run once. Seems tricky to debug.

Reply:
"Record the behavior that will be used then on" as in any operations run as part of the initialization should not be part of the graph (so nothing to do with control dependencies to ensure program order). For example, consider:

```python
v = None
@tf.defun
def f(x):
  global v
  if v is None:
    a = tf.random_uniform(...)
    b = tf.exp(a)
    v = tf.Variable(a)
  return x + v
```

We don't want the operations to initialize the variable (`tf.random_uniform()`, `tf.exp()`) to be included in the graph defined by the trace.  So the proposal is to run `f` twice. The first time it will create the variables and the second time it will generate the graph that will be used in subsequent calls to (the decorated) `f`.  Does that clarify the message? Though, I see your concern, and am very open to suggestions.

Reply:
Thanks for the clarification, @asimshankar . Wouldn't it be cleaner to just require variables to be created separately, as in Keras (where variables are created in `build()`, and computations are performed in `call()`)? This way you wouldn't have to handle the complexity of dealing with variables created within `f()`. Perhaps there's a great use case for this kind of code, but I don't see it, and it seems like a lot of extra complexity (not just in the implementation for the TensorFlow team, but also extra cognitive load for anyone using `defun()` or learning TF).

Reply:
Here's a small [python program](https://gist.github.com/ageron/1a92faa5e72134f49e6bcbcbeecfe8af) that shows different ways to use `@tf.function` (I prefer that name rather than `@tf.defun`) and `@tf.method`.  The first one is nice and clean, and the others are gradually uglier. Only the ugliest solution requires creating some TF state within the function decorated by `@tf.function`.
I think we should just forbid creating any TF state within a function or method decorated by `tf.function` or `tf.method`. But of course there might be use cases I haven't thought of (`python -m this | grep Special`).

Reply:
We're using `@tf.function` now, and there is no need for `@tf.method` as the one decorator suffices. (The RFC is updated with that). So thanks for nudging in that direction :)

Regarding forbidding creation of state - the nice and clean examples you provided will (and should be) encouraged. In fact, the documentation for `tf.function` does encourage state creation within the body of the decorated function.

From a pragmatic standpoint: This caveat about creating state has made it possible to wrap large blocks of existing third-party/user code/libraries in a `tf.function`. For example, the `__call__` method of Keras layers create state on the first invocation (since they invoke `build()`). So, allowing state creation allows things like:

```python
m = <some tf.keras.Model>
m.call = tf.function(m.call) # Use graph execution for the Keras model
m(<some inputs>)
```

without having to redesign the implementation of  Keras.

So for now I'd keep this (for a preview release) and reconsider as we get some more mileage before the final 2.0 release.

Once again, thanks a bunch for your comments.